_posts/2017-12-07-recursion.md

```diff
@@ -0,0 +1,86 @@
+---
+layout: post
+name: 2017-12-07-recursion
+title: Recursion
+date: 2017-12-03 07:00:00 +00:00
+author: Christian Panadero Martinez
+image:
+   src: /assets/img/custom/blog/2017-12-07-recursion.jpg
+tags:
+- functional programming
+- Scala
+- recursion
+---
+
+Recursion is a technique that allows us to break down a problem into smaller pieces. This technique allows us to remove some local <a href="https://codurance.com/2017/11/02/side-effects/">side effects</a> that we perform while writing looping structures and also makes our code more expressive and readable. In this post we will see why it is a very useful technique in functional programming and how it can help us.
+
+<h3>Functional loops</h3>
+In the function that we wrote in the previous post:
+
+<pre class="prettyprint"><code>def sumIntsTo(i: Int) = {
+  var result = 0
+  0 to i foreach((i) => result = result + i)
+  result
+}</code></pre>
+
+You could notice that there is a <a href="https://codurance.com/2017/11/02/side-effects/">side effect</a> in that imperative-like loop. We are assigning the result of the computation to an accumulator because <span style="padding:0;" class="prettyprint"><code>foreach</code></span> is a function that returns Unit. Or in other words, it does not return anything, so we don't have any other choices. That foreach method therefore is a statement. We say that an statement is a block of code that can be executed but not reduced. On the other hand, we say that a block of code that can be reduced is an expression.
+
+<h3>Head recursion</h3>
+The first technique that we are going to use is called head recursion. Let's translate the previous function to a head recursive function:
+
+<pre class="prettyprint"><code>def sumIntsToRecursive(i: Int) : Int =
+  if (i == 0) 0 else i + sumIntsToRecursive(i - 1)</code></pre>
+
+This function does not have any <a href="https://codurance.com/2017/11/02/side-effects/">side effects</a>. Internally it calculates a sum until we reach the base case which is 0. Here is the stack trace for <span style="padding:0;" class="prettyprint"><code>sumIntsToRecursive(5)</code></span>:
```

Comment:
If you've linked to the side effects post before, I don't think you need to link to it again.

Reply:
It is just to use it a dictionary. Some blogs use this technique to increase the SEO. Not sure if ti works tho